### ISSUE TITLE
Import Protobuf types from other packages

### ISSUE THREAD

**Issue by dmdashenkov** [dart-lang/protobuf#295]
Assuming we have 2 Dart packages. 
Package `A` is a library. It defines Protobuf types and Dart utilities to work with them.
Package `B` depends on `A`. It also defines Protobuf types, which depend on types from `A`.

When trying to build `B`, we cannot avoid re-generating Dart code for Protobuf types is `A`. If we don't generate `A` types again, `B` types are generated with relative imports for `B` types and do not compile.
Is there a way to make the plugin use absolute import and configure the mapping of imported types to the Dart packages which contain them?

**Comment by sigurdm** [comment]
No, I don't think we support such a thing currently.

Are you in control of both A and B? maybe you could gather all the proto files in a shared package C?

**Comment by dmdashenkov** [comment]
For us, `A` is a library and `B` uses it. We control `A` but not `B`.
As the last resort, we have an option of manually substituting import statements. But I was hoping that, maybe, there is already a solution for that.

**Comment by sigurdm** [comment]
No unfortunately we do not have a thing like that.

How would such an option work? Would it be an option in the .proto file or on the protoc invocation?
I don't think proto `import`s can have options

**Comment by dmdashenkov** [comment]
I'd expect there to be a file-level Protobuf option. Something like 
```proto
option dart_package = 'helpful_library';
```
Now when generating code, a protoc compiler has access to the descriptor of imported types (through `CodeGeneratorRequest.proto_file`). So, when generating Dart code it is possible to scan for this option and use a package-based import instead of a path-based one.

Maybe, you'll have a better idea. The main disadvantage of this solution is that it'd require changing the standard library types (or hardcoding them). In general, this works only if the upstream types are declared with the option.

**Comment by dmdashenkov** [comment]
This is similar to other options such as `java_pacakge` and `go_package`.

**Comment by sigurdm** [comment]
That would work I think - but would have to go into package A.

**Comment by sigurdm** [comment]
Ah - I read again - you control package A. Then this is the right thing to do.

**Comment by sigurdm** [comment]
@dmdashenkov would https://github.com/dart-lang/protobuf/pull/298 work for your use case?

**Comment by dmdashenkov** [comment]
Yes, that PR should solve my problem, as far as I can see. Thanks.

**Comment by dmdashenkov** [comment]
Hi. Any updates on the feature status?

**Comment by sigurdm** [comment]
Sorry for the silence.
I am trying to get a proper option into protobuf itself. (So you would no longer need to import dart_options) it is taking some time.
Hopefully I'll manage, it seems the right way to go.

**Comment by sigurdm** [comment]
Sorry - this is currently only tracked on an internal PR.
It would appear as   `optional string dart_package = 48;` in descriptor.proto

**Comment by dmdashenkov** [comment]
Hi. Are there any updates on this issue?

**Comment by sigurdm** [comment]
No unfortunately I had some trouble landing this in the google internal protobuf library, and got swept away with other priorities.

**Comment by dave** [comment]
Hey is this going to be back on your radar any time soon? I've been trying to find a workaround for a couple of days now, and without this fix it seems my only option is to bundle all my code together in one package instead of splitting it up. 

It sounds like we're waiting for an internal google change, but if there's anything I can help with to get it moving, let me know - I'm happy to help...

**Comment by dave** [comment]
Just another quick ping on this issue... It would be really great to get it resolved. There's no convenient work-around unfortunately...

**Comment by eldrigo** [comment]
any updates on this?

**Comment by oklimberg** [comment]
Are there any news on this?

**Comment by loeffel-io** [comment]
highly needed
